# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: netwerk/protocol/http/EarlyHintRegistrar.cpp
# Commit: 7cdf3cef2773
# Full Hash: 7cdf3cef27736a14fd8f9db58522da37e103d984
# Author: Manuel Bucher <manuel@mozilla.com>
# Date: 2022-12-02 21:26:33
# Regressor Bug: 1771867
# File Overlap Count: 1
# Description:
#   Bug 1771867 - Early Hints Phase 2 - Part 3: Add mechanism to notify EarlyHintPreloader that HttpChannelParent is ready to complete the redirect r=necko-reviewers,kershaw,valentin
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D161178
# ==============================================================================

diff -r 5090eae24a5c -r 7cdf3cef2773 netwerk/protocol/http/EarlyHintRegistrar.cpp
--- /dev/null	Thu Jan 01 00:00:00 1970 +0000
+++ b/netwerk/protocol/http/EarlyHintRegistrar.cpp	Fri Dec 02 09:45:24 2022 +0000
@@ -0,0 +1,68 @@
+/* -*- Mode: C++; tab-width: 8; indent-tabs-mode: nil; c-basic-offset: 2 -*- */
+/* vim: set ts=8 sts=2 et sw=2 tw=80: */
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#include "EarlyHintRegistrar.h"
+
+#include "EarlyHintPreloader.h"
+#include "mozilla/ClearOnShutdown.h"
+#include "mozilla/StaticPtr.h"
+#include "nsXULAppAPI.h"
+
+namespace {
+mozilla::StaticRefPtr<mozilla::net::EarlyHintRegistrar> gSingleton;
+}  // namespace
+
+namespace mozilla::net {
+
+EarlyHintRegistrar::EarlyHintRegistrar() {
+  // EarlyHintRegistrar is a main-thread-only object.
+  // All the operations should be run on main thread.
+  // It should be used on chrome process only.
+  MOZ_ASSERT(XRE_IsParentProcess());
+  MOZ_ASSERT(NS_IsMainThread());
+}
+
+EarlyHintRegistrar::~EarlyHintRegistrar() { MOZ_ASSERT(NS_IsMainThread()); }
+
+// static
+already_AddRefed<EarlyHintRegistrar> EarlyHintRegistrar::GetOrCreate() {
+  if (!gSingleton) {
+    gSingleton = new EarlyHintRegistrar();
+    mozilla::ClearOnShutdown(&gSingleton);
+  }
+  return do_AddRef(gSingleton);
+}
+
+void EarlyHintRegistrar::DeleteEntry(uint64_t aEarlyHintPreloaderId) {
+  MOZ_ASSERT(NS_IsMainThread());
+
+  mEarlyHint.Remove(aEarlyHintPreloaderId);
+}
+
+void EarlyHintRegistrar::RegisterEarlyHint(uint64_t aEarlyHintPreloaderId,
+                                           EarlyHintPreloader* aEhp) {
+  MOZ_ASSERT(NS_IsMainThread());
+  MOZ_ASSERT(aEhp);
+
+  mEarlyHint.InsertOrUpdate(aEarlyHintPreloaderId, RefPtr{aEhp});
+}
+
+bool EarlyHintRegistrar::LinkParentChannel(uint64_t aEarlyHintPreloaderId,
+                                           nsIParentChannel* aParent,
+                                           uint64_t aChannelId) {
+  MOZ_ASSERT(NS_IsMainThread());
+  MOZ_ASSERT(aParent);
+
+  RefPtr<EarlyHintPreloader> ehp;
+  bool found = mEarlyHint.Get(aEarlyHintPreloaderId, getter_AddRefs(ehp));
+  if (ehp) {
+    ehp->OnParentReady(aParent, aChannelId);
+  }
+  MOZ_ASSERT(ehp || !found);
+  return found;
+}
+
+}  // namespace mozilla::net